core/src/main/scala/scala/pickling/generator/algorithms.scala

```diff
@@ -0,0 +1,90 @@
+package scala.pickling
+package generator
+
+import scala.reflect.api.Universe
+
+// TODO - These need logging messages.
+
+/** An interface so we can pass logging to these algorithms at runtime/during testing. */
+trait AlgorithmLogger {
+  def warn(msg: String): Unit
+  def debug(msg: String): Unit
+  def error(msg: String): Unit
+  def abort(msg: String): Nothing
+}
+
+/** Represents the result of an algorithm call. */
+sealed trait AlgorithmResult {
+  def join(other: => AlgorithmResult): AlgorithmResult
+  def map(f: PickleUnpickleImplementation => PickleUnpickleImplementation): AlgorithmResult =
+     this match {
+       case AlgorithmSucccess(i) => AlgorithmSucccess(f(i))
+       case x => x
+     }
+}
+final case class AlgorithmSucccess(impl: PickleUnpickleImplementation) extends AlgorithmResult {
+  def join(other: => AlgorithmResult): AlgorithmResult = this
+}
+/** A list of reasons why an algorithm failued to run. */
+final case class AlgorithmFailure(reasons: List[String]) extends AlgorithmResult {
+  def join(other: => AlgorithmResult): AlgorithmResult =
+    other match {
+      case x: AlgorithmSucccess => x
+      case AlgorithmFailure(rs) => AlgorithmFailure(reasons ++ rs)
+    }
+}
+object AlgorithmFailure {
+  def apply(reason: String): AlgorithmFailure = AlgorithmFailure(List(reason))
+}
+/** An abstract implementation of a pickling generation algorithm.
+  *
+  *
+  * TODO - Do we even need an interfaace?
+  */
+trait PicklingAlgorithm {
+  /**
+   * Attempts to construct pickling logic for a given type.
+   */
+  def generate(tpe: IrClass, logger: AlgorithmLogger): AlgorithmResult
```

Comment:
It's literally meant to do that.  That is the GOAL of this behavior.  When you fail to generate a pickler, you want to know why.  I'm purposefully showing all the algorithms and why `$tpe` doesn't fit any of them.  So no, I don't think `isSupported` is a good idea.

I think I need to show some example error messages so you get a feel for what it means, but basically if we fail to create a pickler, I'd like to know why.

What might make sense is `isOnlyApplicableAlgoirthm` so you can streamline errors.  From what I've seen so far, we don't even know if that method could ever return true.